core/tauri/src/image.rs

```diff
@@ -0,0 +1,205 @@
+  ///
+  /// Only `ico` and `png` are supported (based on activated feature flag).
+  #[cfg(any(feature = "image-ico", feature = "image-png"))]
+  #[cfg_attr(docsrs, doc(cfg(any(feature = "image-ico", feature = "image-png"))))]
+  pub fn from_bytes(bytes: &[u8]) -> std::io::Result<Self> {
+    let extension = infer::get(bytes)
+      .expect("could not determine icon extension")
+      .extension();
+
+    match extension {
+      #[cfg(feature = "image-ico")]
+      "ico" => Self::from_ico_bytes(bytes),
+      #[cfg(feature = "image-png")]
+      "png" => Self::from_png_bytes(bytes),
+      _ => {
+        #[cfg(all(feature = "image-png", feature = "image-ico"))]
+        let supported = "'ico' or 'png'";
+        #[cfg(all(
+          not(all(feature = "image-ico", feature = "image-png")),
+          feature = "image-ico"
+        ))]
+        let supported = "'ico'";
+        #[cfg(all(
+          not(all(feature = "image-ico", feature = "image-png")),
+          feature = "image-png"
+        ))]
+        let supported = "'png'";
+        #[cfg(not(any(feature = "image-png", feature = "image-ico")))]
+        let supported = "''";
+
+        Err(Error::new(
+          ErrorKind::InvalidInput,
+          format!("Unexpected image format, expected {supported}, found '{extension}'"),
+        ))
+      }
+    }
+  }
+
+  /// Creates a new image using the provided path.
+  ///
+  /// Only `ico` and `png` are supported (based on activated feature flag).
+  #[cfg(any(feature = "image-ico", feature = "image-png"))]
+  #[cfg_attr(docsrs, doc(cfg(any(feature = "image-ico", feature = "image-png"))))]
+  pub fn from_path<P: AsRef<std::path::Path>>(path: P) -> std::io::Result<Self> {
+    let bytes = std::fs::read(path)?;
+    Self::from_bytes(&bytes)
+  }
+
+  /// Returns the RGBA data for this image, in row-major order from top to bottom.
+  pub fn rgba(&'a self) -> &'a [u8] {
+    &self.rgba
+  }
+
+  /// Returns the width of this image.
+  pub fn width(&self) -> u32 {
+    self.width
+  }
+
+  /// Returns the height of this image.
+  pub fn height(&self) -> u32 {
+    self.height
+  }
+}
+
+impl<'a> From<Image<'a>> for crate::runtime::Icon<'a> {
+  fn from(img: Image<'a>) -> Self {
+    Self {
+      rgba: img.rgba,
+      width: img.width,
+      height: img.height,
+    }
+  }
+}
+
+#[cfg(desktop)]
+impl TryFrom<Image<'_>> for muda::Icon {
+  type Error = crate::Error;
+
+  fn try_from(img: Image<'_>) -> Result<Self, Self::Error> {
+    muda::Icon::from_rgba(img.rgba.to_vec(), img.width, img.height).map_err(Into::into)
+  }
+}
+
+#[cfg(all(desktop, feature = "tray-icon"))]
+impl TryFrom<Image<'_>> for tray_icon::Icon {
+  type Error = crate::Error;
+
+  fn try_from(img: Image<'_>) -> Result<Self, Self::Error> {
+    tray_icon::Icon::from_rgba(img.rgba.to_vec(), img.width, img.height).map_err(Into::into)
+  }
+}
+
+#[cfg(desktop)]
+#[derive(serde::Deserialize)]
+#[serde(untagged)]
+pub enum JsIcon<'a> {
```

Comment:
why isn't this called JsImage?

Reply:
Have no idea tbh 😅, I will open another PR for JS implementation